### PR TITLE
`Core`: Add course mailing campaigns

### LINKED ISSUE #1733: Email all course participants
Provide a way to send an email to all participants of a course.

### PR DESCRIPTION
**Pull request by Mtze**
## Summary

Adds a per-course **Mailing** feature: a sidebar item that opens a campaign overview and a composer for emailing students by phase and status, with saved, reusable, trackable campaigns.

## Details

**Backend** (`servers/core/courseMailing`)

- New course-scoped module under `/api/courses/:uuid/mail-campaigns` (create/list/get/update/copy/delete, recipient preview, test-send, send, resend-failed). Editors can draft; only lecturers/admins can send.
- Migration `0026` adds `mail_campaign` + `mail_campaign_recipient` (status enums). Recipient rows store only a snapshot email plus the per-recipient outcome; student names join live to `student` (no duplicated student data).
- Async sender: bounded worker pool, atomic send-guard (`409` on concurrent send), zero-recipient guard (`422`), CC/BCC sent once (not per student), resend-to-failed re-resolving placeholders live, and crash reconciliation on startup. The target phase is validated against the campaign's own course (prevents cross-course targeting). Reuses the existing `mailing` SMTP/placeholder code via thin exported wrappers.

**Frontend** (`clients/core`)

- New `Mailing` sidebar entry + routes; an overview table (status badges, recipient counts, metadata, row actions) and a composer (phase + multi-status select, shared `EmailTemplateEditor` with placeholders, recipient preview, per-campaign reply-to/CC/BCC overrides, test-send, send-with-confirmation). Replaces the unused `MailingConfigSidebar`.

**Testing & docs**

- Go `testcontainers` suite (CRUD, copy, send edge cases, concurrency `409`, zero-recipient, resend, CC/BCC-once, cross-course rejection, role gating).
- Playwright spec plus a `mailpit` mail catcher wired into `docker-compose.e2e.yml`; the spec asserts delivery.
- `docs/user/mailing.md` extended.

## Reason / Link to issue

Provides a way to email the participants of a course. Closes #1733.

## How to Test

1. `cd servers/core && go test ./courseMailing/...`
2. `make lint && make test`
3. `make test-e2e` — full suite including the mailing spec (asserts delivery via `mailpit`).
4. Manual: open a course → **Mailing** → New mail → pick a phase + statuses → preview recipients → Test send → Send.

## PR Checklist

- [x] Tested locally or on the dev environment
- [x] Code is clean, readable, and documented
- [x] Tests added or updated (if needed)
- [ ] Screenshots attached for UI changes (if any)
- [x] Documentation updated (if relevant)

## Summary by CodeRabbit

* **New Features**
  * Added course-wide mailing campaigns for creating, editing, copying, previewing, testing, sending, and tracking emails.
  * Added recipient selection, status tracking, delivery results, and retrying failed recipients.
  * Added campaign status indicators, recipient previews, confirmation dialogs, and campaign management actions.
  * Added Mailing to the course management navigation.
  * Restricted campaign management and sending actions based on course permissions.
* **Documentation**
  * Added guidance for configuring and using course mailing campaigns.

### CODE REVIEW

File: clients/core/src/managementConsole/courseMailing/components/MailItemListEditor.tsx [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+            onChange={(e) => update(index, { email: e.target.value })}
+          />
+          <Input
+            placeholder='name (optional)'
+            value={item.name}
+            onChange={(e) => update(index, { name: e.target.value })}
+          />
+          <Button variant='ghost' size='icon' type='button' onClick={() => remove(index)}>
+            <Trash2 className='h-4 w-4' />
+          </Button>
```
[the remark this answers is not in the document]
Comment (Mtze) [reply clients/core/src/managementConsole/courseMailing/components/MailItemListEditor.tsx]: Fixed in 56c421af - added `aria-label={`Remove ${label} recipient ${index + 1}`}` to the icon-only remove button.

File: clients/core/src/managementConsole/courseMailing/CourseMailingComposePage.tsx [hunk 2]
```diff
@@ -0,0 +1,421 @@
...
+        </Button>
+      </div>
+
+      <SendConfirmationDialog
+        isOpen={sendDialogOpen}
+        onClose={() => setSendDialogOpen(false)}
+        onConfirm={handleSend}
+        recipientCount={recipientPreview?.count ?? 0}
+        isPending={isSending}
+      />
```
[the remark this answers is not in the document]
Comment (Mtze) [reply clients/core/src/managementConsole/courseMailing/CourseMailingComposePage.tsx:432]: Fixed in 56c421af - the compose page now passes an `isCountKnown` flag (derived from the preview query's `isLoading`/`isError`) to `SendConfirmationDialog`, which shows "Resolving the recipient list..." and keeps Send disabled until the count actually resolves.

File: clients/core/src/managementConsole/courseMailing/hooks/useCourseMailingCampaigns.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+export const useGetRecipientPreview = (
+  courseID: string,
+  campaignID: string | undefined,
+  enabled: boolean,
+) =>
+  useQuery({
+    queryKey: ['mailCampaignRecipientPreview', courseID, campaignID ?? ''],
+    queryFn: () => getRecipientPreview(courseID, campaignID ?? ''),
+    enabled: enabled && !!campaignID,
+  })
```
[the remark this answers is not in the document]
Comment (Mtze) [reply clients/core/src/managementConsole/courseMailing/hooks/useCourseMailingCampaigns.ts:57]: Fixed in 56c421af - `useUpdateMailCampaign` now also invalidates the `['mailCampaignRecipientPreview', courseID, campaignID]` key on success, so the preview and the send-dialog count refresh after a targeting change.

File: docker-compose.e2e.yml [hunk 4]
```diff
@@ -753,6 +760,18 @@ services:
...
+  mailpit:
+    image: axllent/mailpit:v1.21.8
+    container_name: prompt-e2e-mailpit
+    environment:
+      - MP_SMTP_AUTH_ACCEPT_ANY=true
+      - MP_SMTP_AUTH_ALLOW_INSECURE=true
+    # HTTP UI/API published on 18025 so host-run Playwright can reach it too.
+    ports:
+      - "18025:8025"
+
```
[the remark this answers is not in the document]
Comment (Mtze) [reply docker-compose.e2e.yml:853]: Declined - the mailpit service has no healthcheck; `server-core` depends on it with `condition: service_started`, and the runner polls the app's `/api/hello` for readiness. The full e2e suite (173/173) passes with mailpit-asserted delivery, so there's nothing to gate on `readyz` here.

File: docs/user/mailing.md [hunk 5]
```diff
@@ -101,6 +101,52 @@ Students who do not have a status assigned, when sending out the mails, will not
 
 ---
 
+## 📨 Course Mailing Campaigns
+
+Beyond the per-phase status mails, every course has a **Mailing** entry in the course sidebar for
+composing, saving, and sending reusable email campaigns to a chosen group of students.
```
[the remark this answers is not in the document]
Comment (Mtze) [reply docs/user/mailing.md:107]: Fixed in 56c421af - but note the mailing *configuration* now lives under the **Settings** entry (rendered by `CourseSettingsPage`), not a standalone "Mailing Settings" item; the new **Mailing** entry is for campaigns. Updated the earlier instruction to point at **Settings** rather than renaming it to "Mailing".

File: e2e/src/pages/CourseMailingPage.ts [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    await this.newMailButton.click()
+    await expect(this.page.getByRole('heading', { name: 'New Campaign' })).toBeVisible()
+  }
+
+  async fillDetails(input: CampaignInput) {
+    await this.page.getByLabel('Name', { exact: true }).fill(input.name)
+
+    // Course phase (Radix Select → combobox trigger).
+    await this.page.getByRole('combobox').click()
+    await this.page.getByRole('option', { name: input.phaseName }).click()
```
[the remark this answers is not in the document]
Comment (Mtze) [reply e2e/src/pages/CourseMailingPage.ts:41]: Declined - the e2e suite passes (173/173) with the current selector, so it resolves the MultiSelect reliably in practice. Left as-is to avoid churn on a green selector.

File: servers/core/courseMailing/router.go [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+func bindCampaignRequest(c *gin.Context) (courseMailingDTO.MailCampaignRequest, error) {
+	var req courseMailingDTO.MailCampaignRequest
+	if err := c.BindJSON(&req); err != nil {
+		return courseMailingDTO.MailCampaignRequest{}, err
+	}
+	if err := validateCampaignRequest(req); err != nil {
+		return courseMailingDTO.MailCampaignRequest{}, err
+	}
+	return req, nil
+}
```
[the remark this answers is not in the document]
Comment (Mtze) [reply servers/core/courseMailing/router.go:328]: Fixed in 56c421af - `bindCampaignRequest` now uses `c.ShouldBindJSON`, so the handler owns the error mapping instead of `BindJSON` aborting with its own 400 and double-writing.

File: servers/core/courseMailing/router.go [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+		handleError(c, http.StatusConflict, err)
+	case errors.Is(err, ErrNoRecipients):
+		handleError(c, http.StatusUnprocessableEntity, err)
+	case errors.Is(err, ErrValidation):
+		handleError(c, http.StatusBadRequest, err)
+	default:
+		log.Error(err)
+		handleError(c, http.StatusInternalServerError, err)
+	}
+}
```
[the remark this answers is not in the document]
Comment (Mtze) [reply servers/core/courseMailing/router.go:349]: Fixed in 56c421af - the default branch now logs the wrapped error and returns a fixed `"internal server error"` message, so pgx/table details no longer reach the client.

File: servers/core/courseMailing/send.go [hunk 9]
```diff
@@ -0,0 +1,508 @@
...
+		body:        base.Body,
+		items:       items,
+		settings:    settings,
+		courseInfo:  courseInfo,
+		actor:       actor,
+		setSentMeta: true,
+	}
+	runSendAsync(func() { s.runCampaignSend(job) })
+
+	return len(rows), nil
```
[the remark this answers is not in the document]
Comment (Mtze) [reply servers/core/courseMailing/send.go]: Fixed in 56c421af - `SendCampaign` now returns `len(items)` (the number actually dispatched), matching `ResendFailed`.

File: servers/core/courseMailing/send.go [hunk 10]
```diff
@@ -0,0 +1,508 @@
...
+		CampaignID:   job.campaignID,
+		ErrorMessage: "send did not complete",
+	}); err != nil {
+		log.WithError(err).WithField("campaignID", job.campaignID).Warn("failed to fail leftover pending recipients")
+	}
+
+	if err := s.finalizeSend(statusCtx, job); err != nil {
+		log.WithError(err).WithField("campaignID", job.campaignID).Error("failed to finalize campaign send")
+	}
+}
```
[the remark this answers is not in the document]
Comment (Mtze) [reply servers/core/courseMailing/send.go:398]: Declined - `mailing.SendCourseMail` is shared across the mailing module and setting per-dial SMTP deadlines there is out of scope for this PR. The send runs under a 30-minute per-campaign context, and `FailStalePendingRecipients` flips any leftover `pending` rows to `failed` on completion/timeout, so a slow SMTP server can't leave the campaign stuck.

File: servers/core/courseMailing/send.go [hunk 11]
```diff
@@ -0,0 +1,508 @@
...
+	statusCtx := context.WithoutCancel(ctx)
+	if item.email == "" {
+		s.setRecipientFailed(statusCtx, item.recipientID, "missing email address")
+		return
+	}
+	if err := sendMailFn(settings, item.email, item.subject, item.body); err != nil {
+		log.WithError(err).WithField("email", item.email).Warn("failed to send campaign mail to recipient")
+		s.setRecipientFailed(statusCtx, item.recipientID, err.Error())
+		return
+	}
```
[the remark this answers is not in the document]
Comment (Mtze) [reply servers/core/courseMailing/send.go:411]: Fixed in 56c421af - the failure log now uses `WithField("recipientID", item.recipientID)` instead of the address; the email is still stored on the recipient row for the UI.

File: servers/core/courseMailing/validation.go [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+		return fmt.Errorf("%w: campaign name is required", ErrValidation)
+	}
+	if _, err := validateStatusValues(req.TargetPassStatuses); err != nil {
+		return err
+	}
+	if req.ReplyToOverride != nil && req.ReplyToOverride.Email == "" {
+		return fmt.Errorf("%w: reply-to override requires an email address", ErrValidation)
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (Mtze) [reply servers/core/courseMailing/validation.go:52]: Fixed in 56c421af - `validateCampaignRequest` now validates the reply-to override and every CC/BCC entry with `net/mail.ParseAddress` (blank field-array rows are ignored, matching `parseAddresses`). Added a regression test.

File: servers/core/courseMailing/validation.go [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+		return fmt.Errorf("%w: the subject must contain only ASCII characters", ErrValidation)
+	}
+	if strings.TrimSpace(base.Body) == "" {
+		return fmt.Errorf("%w: a message body is required before sending", ErrValidation)
+	}
+	if !base.TargetCoursePhaseID.Valid {
+		return fmt.Errorf("%w: a target course phase must be selected before sending", ErrValidation)
+	}
+	return nil
+}
```
[the remark this answers is not in the document]
Comment (Mtze) [reply servers/core/courseMailing/validation.go:81]: Fixed in 56c421af - `buildBaseMailHeader` writes `Subject: %s\r\n` raw, and ASCII allows CR/LF, so `validateCampaignForTest` now rejects any subject containing line breaks to prevent header injection. Added a regression test.

File: servers/core/db/migration/0028_course_mail_campaigns.up.sql [hunk 14]
```diff
@@ -0,0 +1,44 @@
+-- Migration: Add course mail campaign tables
+
+CREATE TYPE mail_campaign_status AS ENUM ('draft', 'sending', 'sent', 'partially_failed', 'failed');
+CREATE TYPE mail_recipient_status AS ENUM ('pending', 'sent', 'failed');
```
[the remark this answers is not in the document]
Comment (Mtze) [reply servers/core/db/migration/0028_course_mail_campaigns.up.sql:4]: Declined - this repo uses an up-only migration convention (no `.down.sql` files anywhere under `db/migration`), so the new migration correctly ships without a down file.

File: servers/core/docs/docs.go [hunk 15]
```diff
@@ -2341,6 +2341,357 @@ const docTemplate = `{
...
+                    "200": {
+                        "description": "OK",
+                        "schema": {
+                            "type": "array",
+                            "items": {
+                                "$ref": "#/definitions/courseMailingDTO.MailCampaign"
+                            }
+                        }
+                    }
+                }
```
[the remark this answers is not in the document]
Comment (Mtze) [reply servers/core/docs/docs.go:2703]: Fixed in 56c421af - added `@Failure` annotations (400/404/409/422/500 as applicable) to every mail-campaign handler and regenerated `docs/` via `scripts/generate-api-spec.sh core`.

File: servers/core/courseMailing/service.go [hunk 16]
```diff
@@ -0,0 +1,393 @@
...
+	if err := s.validatePhaseBelongsToCourse(ctx, base.CourseID, phaseID); err != nil {
+		return uuid.UUID{}, err
+	}
+	return phaseID, nil
+}
+
+func (s *CourseMailingService) CreateCampaign(ctx context.Context, courseID uuid.UUID, actor courseMailingDTO.Actor, req courseMailingDTO.MailCampaignRequest) (db.MailCampaign, error) {
+	if err := s.validateTargetPhase(ctx, courseID, req.TargetCoursePhaseID); err != nil {
+		return db.MailCampaign{}, err
+	}
```
[the remark this answers is not in the document]
Comment (Mtze) [reply servers/core/courseMailing/service.go:157]: Fixed in 56c421af - `CreateCampaign` now calls `validateCampaignRequest` at the top, so invalid campaigns are rejected even when the service is called directly (e.g. in tests). Added a service-boundary test.

File: servers/core/courseMailing/send.go [hunk 17]
```diff
@@ -0,0 +1,508 @@
...
+// sendArchiveCopy sends a single copy to the campaign's CC/BCC recipients so they
+// are not copied once per student.
+func (s *CourseMailingService) sendArchiveCopy(ctx context.Context, job campaignSendJob) {
+	if len(job.settings.CC) == 0 && len(job.settings.BCC) == 0 {
+		return
+	}
+	placeholders := mailing.StatusPlaceholderValues(job.courseInfo.CourseName, job.courseInfo.CourseStartDate, job.courseInfo.CourseEndDate, db.GetParticipantMailingInformationRow{})
+	subject := mailing.ReplacePlaceholders(job.subject, placeholders)
+	body := mailing.ReplacePlaceholders(job.body, placeholders)
+	if err := sendMailFn(job.settings, job.settings.ReplyTo.Address, subject, body); err != nil {
```
[the remark this answers is not in the document]
Comment (Mtze) [reply servers/core/courseMailing/send.go]: Fixed in 56c421af - the resolved `{{coursePhaseLink}}` is now carried on the send job and injected into the archive-copy placeholder map, so the CC/BCC copy renders the same link as the student mails. Added a test.

File: servers/core/courseMailing/main.go [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+	db "github.com/prompt-edu/prompt/servers/core/db/sqlc"
+	"github.com/prompt-edu/prompt/servers/core/keycloakTokenVerifier"
+	"github.com/prompt-edu/prompt/servers/core/permissionValidation"
+)
+
+func InitCourseMailingModule(routerGroup *gin.RouterGroup, queries db.Queries, conn *pgxpool.Pool) {
+	CourseMailingServiceSingleton = &CourseMailingService{
+		queries:   queries,
+		conn:      conn,
+		clientURL: sdkUtils.GetEnv("CORE_HOST", "localhost:3000"),
```
[the remark this answers is not in the document]
Comment (Mtze) [reply servers/core/courseMailing/main.go]: Fixed in 56c421af - the `CORE_HOST` default is now `http://localhost:3000`, so `{{coursePhaseLink}}` resolves to an absolute URL in local dev.

File: servers/core/courseMailing/service.go [hunk 19]
```diff
@@ -0,0 +1,393 @@
...
+	existing, err := s.getBase(ctx, courseID, campaignID)
+	if err != nil {
+		return db.MailCampaign{}, err
+	}
+	if existing.Status == db.MailCampaignStatusSending {
+		return db.MailCampaign{}, ErrSendInProgress
+	}
+	if err := s.validateTargetPhase(ctx, courseID, req.TargetCoursePhaseID); err != nil {
+		return db.MailCampaign{}, err
+	}
```
[the remark this answers is not in the document]
Comment (Mtze) [reply servers/core/courseMailing/service.go:201]: Fixed in 56c421af - `UpdateCampaign` now runs `validateCampaignRequest` before writing, matching `CreateCampaign`.

File: clients/core/src/managementConsole/courseMailing/CourseMailingComposePage.tsx [hunk 20]
```diff
@@ -0,0 +1,421 @@
+import { PassStatus, useCourseStore, useGetMailingIsConfigured } from '@tumaet/prompt-shared-state'
```
[the remark this answers is not in the document]
Comment (Mtze) [reply clients/core/src/managementConsole/courseMailing/CourseMailingComposePage.tsx:1]: Fixed in 56c421af - added a `useCanSendMailCampaign` hook (PromptAdmin/PromptLecturer/CourseLecturer) that disables the Send and Test-send buttons for course editors, so they no longer trigger a backend 403.

File: clients/core/src/managementConsole/courseMailing/CourseMailingOverviewPage.tsx [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+  Table,
+  TableBody,
+  TableCell,
+  TableHead,
+  TableHeader,
+  TableRow,
+  useToast,
+} from '@tumaet/prompt-ui-components'
+import { Copy, Pencil, Plus, RefreshCw, Trash2 } from 'lucide-react'
+import { useState } from 'react'
```
[the remark this answers is not in the document]
Comment (Mtze) [reply clients/core/src/managementConsole/courseMailing/CourseMailingOverviewPage.tsx:16]: Fixed in 56c421af - the same `useCanSendMailCampaign` hook now gates the "Resend to failed" row action, so it's hidden from course editors.

File: servers/core/db/migration/0028_course_mail_campaigns.up.sql [hunk 22]
```diff
@@ -0,0 +1,63 @@
...
+
+-- Composite-FK targets so a campaign's target phase and a recipient's participation
+-- can be pinned to the same course, not just validated at the service layer.
+ALTER TABLE course_phase ADD CONSTRAINT unique_course_phase_id_course_id UNIQUE (id, course_id);
+ALTER TABLE course_participation ADD CONSTRAINT unique_course_participation_id_course_id UNIQUE (id, course_id);
```
[the remark this answers is not in the document]
Comment (rappm) [reply servers/core/db/migration/0028_course_mail_campaigns.up.sql:9]: Not acting on this one. golang-migrate's postgres driver wraps every migration in a transaction, so `CREATE INDEX CONCURRENTLY` cannot run here without teaching the runner to execute non-transactional migrations, which is an infrastructure change well outside this PR. The two tables are also small in this domain (a handful of phases per course, one participation row per student), migrations run at startup before the service takes traffic, and 0002/0006/0012/0025 all add unique constraints to existing tables the same way. Worth revisiting separately if we ever want online DDL as a general capability.

File: servers/core/courseMailing/service_test.go [hunk 23]
```diff
@@ -0,0 +1,604 @@
...
+			{Status: db.MailRecipientStatusSent},
+			{Status: db.MailRecipientStatusFailed},
+		}, db.MailCampaignStatusPartiallyFailed},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			require.Equal(t, tt.want, computeCampaignStatus(tt.recipients))
+		})
+	}
+}
```
[the remark this answers is not in the document]
Comment (rappm) [reply servers/core/courseMailing/service_test.go:534]: Fixed in 0849f18c. `runCampaignSend` now returns before `finalizeSend` when `FailStalePendingRecipients` errors, leaving the campaign in `sending` so the reconciliation pass retries it. `computeCampaignStatus` also treats any non-sent recipient (pending included) as unfinished, so leftover pending rows can no longer produce a `sent` campaign on any path.

**Review by rappm — APPROVED** [review]
